### PR TITLE
dialects: (x86) PR10 - RRI Operations

### PR DESCRIPTION
**Pull request by KGrykiel**
Added support for instructions with 1 destination register, one source register and one immediate.

### CODE REVIEW

File: tests/filecheck/dialects/x86/x86_assembly_emission.mlir [hunk 1]
```diff
@@ -84,3 +84,6 @@ x86.mi.xor %0, 2, 8 : (!x86.reg<rax>) -> ()
 // CHECK: xor [rax+8], 2
 x86.mi.mov %0, 2, 8 : (!x86.reg<rax>) -> ()
 // CHECK: mov [rax+8], 2
+
+%rri_imul = x86.rri.imul %1, 2 : (!x86.reg<rdx>) -> !x86.reg<rax>
+// CHECK: imul rax, rdx, 2
```
Comment (superlopuh) [inline tests/filecheck/dialects/x86/x86_assembly_emission.mlir:89]:
```suggestion
// CHECK: imul rax, rdx, 2

```

File: tests/filecheck/dialects/x86/x86_ops.mlir [hunk 2]
```diff
@@ -85,3 +85,6 @@ x86.mi.xor %0, 2, 8 : (!x86.reg<>) -> ()
 // CHECK-NEXT: x86.mi.xor %{{.*}}, 2, 8 : (!x86.reg<>) -> ()
 x86.mi.mov %0, 2, 8 : (!x86.reg<>) -> ()
 // CHECK-NEXT: x86.mi.mov %{{.*}}, 2, 8 : (!x86.reg<>) -> ()
+
+%rri_imul = x86.rri.imul %1, 2 : (!x86.reg<>) -> !x86.reg<>
+// CHECK-NEXT: %{{.*}} = x86.rri.imul %{{.*}}, 2 : (!x86.reg<>) -> !x86.reg<>
```
Comment (superlopuh) [inline tests/filecheck/dialects/x86/x86_ops.mlir:90]:
```suggestion
// CHECK-NEXT: %{{.*}} = x86.rri.imul %{{.*}}, 2 : (!x86.reg<>) -> !x86.reg<>

```

File: xdsl/dialects/x86/ops.py [hunk 3]
```diff
@@ -948,6 +948,67 @@ class MI_MovOp(M_MI_Operation[GeneralRegisterType]):
...
+        r2: Operation | SSAValue,
+        immediate: int | AnyIntegerAttr,
+        *,
+        comment: str | StringAttr | None = None,
+        r1: R1InvT,
+    ):
+        if isinstance(immediate, int):
+            immediate = IntegerAttr(
+                immediate, 32
+            )  # the deault immediate size is 32 bits
```
Comment (superlopuh) [inline xdsl/dialects/x86/ops.py:972]:
```suggestion
            )  # the default immediate size is 32 bits
```

File: xdsl/dialects/x86/ops.py [hunk 4]
```diff
@@ -948,6 +948,67 @@ class MI_MovOp(M_MI_Operation[GeneralRegisterType]):
     name = "x86.mi.mov"
 
 
+class R_RRI_Operation(Generic[R1InvT, R2InvT], IRDLOperation, X86Instruction, ABC):
+    """
+    A base class for x86 operations that have one destination register, one source register and an immediate value.
+    """
+
+    r2 = operand_def(R2InvT)
+    immediate: AnyIntegerAttr = attr_def(AnyIntegerAttr)
```
[the remark this answers is not in the document]
Comment (KGrykiel) [reply xdsl/dialects/x86/ops.py:957]: image
According to the x86 manual immediates can only have 8, 16 or 32 bits (with 32 being the default I believe) with 64 bit immediates only appearing for rare instructions afaik.

[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/x86/ops.py:957]: Ok, feels like we might want to encode that at some point, but not urgently

[the remark this answers is not in the document]
Comment (KGrykiel) [reply xdsl/dialects/x86/ops.py:957]: good to merge then?

**Review by superlopuh — APPROVED** [review]

**Pull request merged**